src/compiler/crystal/semantic/main_visitor.cr

```diff
@@ -779,6 +779,13 @@ module Crystal
     def type_assign(target : InstanceVar, value, node)
       # Check if this is an instance variable initializer
       unless @scope
+        # `InstanceVar` assignment appered in block is not checked
+        # by `Crystal::InstanceVarsInitializerVisitor` because this block
+        # may be passed to a macro. So, it checks here.
+        if current_type.is_a?(Program) || current_type.is_a?(FileModule)
```

Comment:
I was considering, then I think `!current_type.is_a?(InstanceVarContainer)` is not needed because available not `InstanceVarContainer` `current_type` is one of `EnumType` and `LibType` as I know, but both type's body cannot contain assignment (and any expressions). Even if `enum` or `lib` allow to include macro calls in body, assignment is rejected on parse phase, I think.